### PR TITLE
PHP 8 Compatibility

### PR DESCRIPTION
**Pull request by AlexandruGG**
This PR ensures this library is always checked for full PHP 8 compatibility, based on the discussion https://github.com/loophp/collection/discussions/103

- [x] updating CI workflows to run on PHP 8 as well as PHP 7.4
- [x] fixing a couple static analysis issues
- [x] handle usage of `asyncMap` in PHP in the `CollectionSpec`. Due to an issue https://github.com/phpspec/phpspec/issues/1375 with PHPSpec the test fails. Should be fixed soon however
- [x] requiring the PHP 8 jobs to merge in the `settings.yml`

### CODE REVIEW

File: src/Contract/Operation/Applyable.php
Comment (AlexandruGG) [inline src/Contract/Operation/Applyable.php:24]: the interface was out of sync with the [operation](https://github.com/loophp/collection/pull/104/files#diff-79d43bca592b76fa0fb2583de4ea453a994e67ea88ab8d27e403579525c8c834R31)

File: spec/loophp/collection/CollectionSpec.php
Comment (AlexandruGG) [inline spec/loophp/collection/CollectionSpec.php:241]: I've modified this test case to show that an `Error` is thrown for `asyncMap` in PHP > 8.0.

That way the test still passes and we have this visibility

File: .github/settings.yml
Comment (AlexandruGG) [inline .github/settings.yml:17]: I assume these "required to merge" settings will only apply for future PRs after this is merged, right?

File: src/Contract/Operation/Applyable.php
Comment (drupol) [reply src/Contract/Operation/Applyable.php:24]: Maybe we should fix that in another PR, wdyt?

File: spec/loophp/collection/CollectionSpec.php
Comment (drupol) [reply spec/loophp/collection/CollectionSpec.php:241]: I did [this](https://github.com/friends-of-phpspec/phpspec-requires) once upon a time, maybe we should see if it would be a better choice?

File: .github/settings.yml
Comment (drupol) [reply .github/settings.yml:17]: yes :)

File: spec/loophp/collection/CollectionSpec.php
Comment (AlexandruGG) [reply spec/loophp/collection/CollectionSpec.php:241]: that would be nice but even better actually I was going to ask if we should just throw an intentional exception from `asyncMap` itself if it's called in an environment running php 8. What do you think? @drupol 

That way we make it clear it's unsupported for now and we can add this info the [documentation](https://loophp-collection.readthedocs.io/en/latest/pages/api.html#asyncmap) as well

Comment (drupol) [reply spec/loophp/collection/CollectionSpec.php:241]: Yes this is a good idea.

File: src/Contract/Operation/Applyable.php
Comment (AlexandruGG) [reply src/Contract/Operation/Applyable.php:24]: well it's causing an issue with PHPStan in this PR so that's why I updated it here 😄 

It's related to [this error which was ignored](https://github.com/loophp/collection/pull/104/files#diff-22fe68f6ec4412b512a01a6526b11f8ad2f9e9c07ebcd32d018a131037059de7L14-L16). In PHP 8 the parameter name for `mb_strtolower` was changed to from `$str` to `$string`, so there would be x2 errors now depending on which PHP version PHPStan is run. I thought I'd just go ahead and fix it :)

File: spec/loophp/collection/CollectionSpec.php
Comment (AlexandruGG) [reply spec/loophp/collection/CollectionSpec.php]: updated 😁

Comment (drupol) [reply spec/loophp/collection/CollectionSpec.php]: Saw it, seems like it is what we need. Nice!

File: src/Operation/AsyncMap.php
Comment (AlexandruGG) [inline src/Operation/AsyncMap.php:49]: this needs to be inside the class otherwise PHPStan fails when discovering the files and that error cannot be ignored. Won't make a difference for the end user

File: .scrutinizer.yml
Comment (AlexandruGG) [inline .scrutinizer.yml:16]: this is based on [their docs](https://scrutinizer-ci.com/docs/build/multiple_test_environment#use-case-testing-different-versions), but I'm not familiar with this tool at all. I think this will only apply after this PR is merged, right?

File: src/Contract/Operation/Applyable.php
Comment (drupol) [reply src/Contract/Operation/Applyable.php:24]: Ok.

File: .scrutinizer.yml
Comment (drupol) [reply .scrutinizer.yml:16]: Yes :)

File: src/Contract/Operation/Applyable.php
Comment (drupol) [reply src/Contract/Operation/Applyable.php:24]: I made a commit on `master` to fix it, you can rebase your branch if you want.

File: composer.json
Comment (drupol) [inline composer.json:23]: Why this change ?

See: https://github.com/symfony/symfony/pull/36876

File: src/Collection.php
Comment (drupol) [inline src/Collection.php:169]: I think we should normalize this everywhere.

Let's leave it like this today, but this is something I would like to change and normalize everywhere.

**Review by drupol — CHANGES_REQUESTED** [review]
Hi @AlexandruGG,

Very nice work, thanks!

I gave my feedback, let me know what you think.

File: composer.json
Comment (AlexandruGG) [reply composer.json:23]: Lots of heated discussions in that PR 😄 .

My main rationale is the one here: https://github.com/symfony/symfony/pull/36876#issuecomment-631351947. I don't see any good arguments against that comment in the thread. Basically we don't know for example how this library will perform in PHP 8.1, so if we use >= 7.4 then a user installing it on 8.1 might have a bad experience if they find that it's not working.

That being said I don't feel too strongly about this, because there's an argument to be made that users might report issues so you can thus fix them earlier.

**Comment by AlexandruGG** [comment]
Don't merge this yet as I want to see what comes out of the discussion on `asyncMap` https://github.com/loophp/collection/discussions/103. It looks like the library is working in PHP 8, it's just PHPSpec that's failing

**Comment by AlexandruGG** [comment]
@drupol hey, I think we should merge this because it adds the PHP 8 CI jobs and so will ensure any future code changes will be compatible with PHP 8.

It might take a bit to get the phpspec [issue](https://github.com/phpspec/phpspec/issues/1375) merged and released, but at that point we can just update [this test](https://github.com/loophp/collection/pull/104/files#diff-3682f293747f0968dc35d26e4ed1f8d6d059c378d0b81f1100f3dda60c09769cR233-R237).

**Pull request merged**
